lalrpop/src/rust/mod.rs

```diff
@@ -82,32 +92,48 @@ impl<'me, W: Write> RustWrite<W> {
                 if !first && session.emit_whitespace {
                     write!(self.write, " ")?;
                 }
-                write!(self.write, "{},", i)?;
+                write!(self.write, "{i},")?;
                 first = false;
             }
         }
         writeln!(self.write)
     }
 
-    pub fn writeln(&mut self, out: &str) -> io::Result<()> {
-        let buf = out.as_bytes();
+    // This function is intentionally not implemented with the `Write` trait
+    // because it is only called using the `rust!` macro, and thus with a `buf`
+    // that always ends with a `\n`.
+    #[doc(hidden)]
+    #[track_caller]
+    pub fn write_fmt(&mut self, args: fmt::Arguments<'_>) -> io::Result<()> {
+        // avoid intermediate allocation for simple format strings
+        let formatted;
+        let buf = match args.as_str() {
+            Some(s) => s.as_bytes(),
+            None => {
+                formatted = fmt::format(args);
+                formatted.as_bytes()
+            }
+        };
 
         // pass empty lines through with no indentation
-        if buf.is_empty() {
-            return self.write.write_all("\n".as_bytes());
+        match buf {
+            [] => return Ok(()),
+            [b'\n'] => return self.write.write_all(b"\n"),
+            [.., b'\n'] => {}
+            _ => unreachable!("rust! macro should always end with a newline"),
```

Comment:
It's a nitpick, but I wonder if `unreachable!` is the right semantics. I think the common usage is that `unreachable` is used when you can prove locally that the codepath will never be taken, for example:

```
if foo.is_none() {
  do_stuff();
  match foo {
    Some(_) => unreachable!()
    None => ...,
  }
}
```

Here this is an invariant that must be ensured by the caller of `rust!`, and it can most probably be broken at some point by some careless or unknowing maintainer. Maybe just a `panic!` instead? (Of course, it's mostly about good practices , because it doesn't change much in practice: this code just panics the same).

Reply:
Yeah, but my point is that it _could be_ called without the trailing `\n` by a careless author in the future, couldn't it? If yes, then I think unreachable isn't the right semantics. I think unreachable indicates code which is, by all means and whatever the caller is passing to this function, can be proven (on paper) to never be triggered, but the compiler isn't smart enough to prove it. However, I don't think anything forbids at compile time to call this function without a trailing `\n`. So it doesn't look like unreachable code to me, but rather it looks like an invariant which is assumed to be enforced throughout the codebase.

Reply:
While it is technically possible for a careless author to call the function without the trailing '\n', it is highly unlikely given that it is an implementation detail, being only public because of the needs of the macro. Someone would have to go out of their way by calling `x.write_fmt(format_args!("..."))` instead of `rust!(x, "...")`.

But we're arguing over semantics here, in practice both just panic at runtime.

Reply:
I think I wouldn't use unreachable even if the method wasn't public. I don't think it's an issue of public symbols versus non public: it's an issue of locality. Exactly like `unsafe`: you can declare your method safe if you can enforce/prove the safety invariants _locally_, notwithstanding the caller. If you depend on your caller satysifying invariants, you have to declare your method unsafe as well. It's just even worse if it's public, because now any user of LALRPOP can break the invariant and potentially see the result of an `unreachable` call.

The LALRPOP developers and maintainers are subject to mistakes as well, come and go, and don't know the all of the codebase in depth. In general the whole point of safety isn't to say oh this is unlikely to happen, but rather try to foresee carefully anything that could happen, even if not very probable. Of course the cost of safety is to be taken into account (verbosity/complexity), but here it's zero.

A good test is: would you feel confident to change your `unreachable` to `unreachable_unchecked`? I know I wouldn't. I don't really care about this specific case, it doesn't change much with respect to safety or failure mode, but it's more that I don't really understand what is the argument for keeping unreachable either. The error message and the underlying intents are still a bit different, if this ever triggers.

Reply:
Reading the (granted, sparse) [documentation of unreachable](https://doc.rust-lang.org/std/macro.unreachable.html), I respectfully disagree. There doesn't seem to be a "user vs library" distinction there, and the examples doesn't seem to include "breaking an arbitrary invariant required by the function" but more about what I describe above (the this code path is actually impossible because I've tested the condition before but the compiler can't prove it). That being said, a bit more research seems to show that your interpretation does exist in the wild as well.

 I won't fight this further though, as the impact is limited in practice, our personal time is valuable, and @youknowone seems to have a different interpretation than me. @DaniPopes I let you do whatever you think is best and then resolve the conversation :slightly_smiling_face:

Reply:
Though, in this specific case, keep in mind that this function is still part of the public API. So even with a laxer interpretation, I still think it's not a great idea to let a possible codepath where a consumer can trigger unrechable exists.

Reply:
I agree my interpretation is definitely out of the examples in the document (and barely fit in the description, possibly just unintended ambiguity)

Reply:
I am going to leave it as `unreachable!` because lalrpop already use unreachable! for logical unreachable code in `fn symbol_table`.